pkg/ingester/ingester.go

```diff
@@ -1928,6 +1906,73 @@ func (i *Ingester) openExistingTSDB(ctx context.Context) error {
 	return nil
 }
 
+func (i *Ingester) getConcurrencyConfig(userCount int) (tsdbOpenConcurrency, walReplayConcurrency int) {
+	tsdbOpenConcurrency = i.cfg.BlocksStorageConfig.TSDB.DeprecatedMaxTSDBOpeningConcurrencyOnStartup
+	walReplayConcurrency = 0
+	// If TSDBConfig.WALReplayConcurrency is set to a positive value, we honor it and ignore value of
+	// TSDB.DeprecatedMaxTSDBOpeningConcurrencyOnStartup, being the latter deprecated.
+	// If TSDBConfig.WALReplayConcurrency is 0, it is ignored, and TSDB.DeprecatedMaxTSDBOpeningConcurrencyOnStartup
+	// determines the number of concurrent processes opening TSDBs.
+	if i.cfg.BlocksStorageConfig.TSDB.WALReplayConcurrency > 0 {
```

Comment:
[nit] The comment above explains what the code does, but this code is quite trivial so not sure there's much value explaining it. Instead, we should explain **why**. I suggest something like this:

```
When WALReplayConcurrency is enabled, we want to ensure the WAL replay at ingester startup
doesn't use more than the configured number of CPU cores. The get best performance both on
single tenant and multi tenant Mimir clusters, we use an heuristic to decide whether it's better to
parallelise the WAL replay of each single TSDB (low number of tenants) or the WAL replay of
multiple TSDBs at the same time (high number of tenants).
```